### PR TITLE
Fix netns leak on container creation and exit code 1 on SIGTERM.

### PR DESCRIPTION
**Pull request by Luap99**
libpod: ensure we are not killed during netns creation

When we are killed during netns setup it will leak the netns path as it
was not commited in the db. This is rather common if you run systemctl
stop on a podman systemd unit. Of course we cannot protect against
SIGKILL but in systemd case we get SIGTERM and we really should not exit
in a critical section like this.

Fixes https://github.com/containers/podman/issues/24044

---

libpod: rework shutdown handler flow

Currently podman run -d can exit 0 if we send SIGTERM during startup
even though the contianer was never started. That just doesn't make any
sense is horribly confusing for a external job manager like systemd.

The original motivation was to exit 0 for the podman.service in commit
https://github.com/containers/podman/commit/ca7376bb115e2d3506a3efe242bde25118f57c39. That does make sense but it should only do so for the
service and only if the server did indeed gracefully shutdown.

So we rework how the exit logic works, do not let the handler perform
the exit. Instead the shutdown package does the exit after all handlers
are run, this solves the issue of ordering. Then we default to exit code
1 like we did before and allow the service exit handler to overwrite the
exit code 0 in case of a graceful shutdown.

---
libpod: remove shutdown.Unregister()

It is never used and needed so let's just remove some dead code.

#### Does this PR introduce a user-facing change?

```release-note
Podman no longer exits 0 on SIGTERM by default.
Fixed a race that could cause podman to leak netns files when it was interrupted during the netns creation.
```

### CODE REVIEW

**Comment by Luap99** [comment]
@mheon @edsantiago PTAL

**Comment by edsantiago** [comment]
LGTM, and I can no longer reproduce the netns leak

File: libpod/shutdown/handler.go [hunk 1]
```diff
@@ -75,6 +83,7 @@ func Start() error {
 			}
 			handlerLock.Unlock()
 			shutdownInhibit.Unlock()
+			os.Exit(exitCode)
```
Comment (mheon) [inline libpod/shutdown/handler.go:86]: Wouldn't this prevent a lot of `defer` functions from running? I know I avoided it deliberately when I wrote this

Comment (Luap99) [reply libpod/shutdown/handler.go:86]: I mean yes but if we exit what do we want to defer here? And most importantly the current handler also just exited so there should not be any functional difference I would say

Comment (mheon) [reply libpod/shutdown/handler.go:86]: IIRC we have a bunch of things running in defer that do things like removing files to clean up after ourselves

Comment (Luap99) [reply libpod/shutdown/handler.go:86]: Sure but this PR doesn't change the behavior here. Critical sections where we cannot leak need to use `shutdown.Inhibit()` which is how I fixed the issue with the netns leak.

**Comment by baude** [comment]
LGTM, but defereing for merge to @mheon given his questions ...

**Comment by edsantiago** [comment]
Saw [the netns leak flake](https://api.cirrus-ci.com/v1/artifact/task/5325582806810624/html/sys-podman-debian-13-root-host-sqlite.log.html#t--00281) today, almost wept in despair, then remembered that this PR hasn't merged yet.

**Comment by mheon** [comment]
/lgtm
I am hesitant but I cannot remember why we didn't want to use exit so I'll merge and hopefully we don't find out the hard way later

**Comment by Luap99** [comment]
>  I am hesitant but I cannot remember why we didn't want to use exit so I'll merge and hopefully we don't find out the hard way later

The old code did use exit() there is no functional change in that regard.

**Comment by mheon** [comment]
LGTM

**Pull request merged**